### PR TITLE
Pre post conditions

### PR DESCRIPTION
**Pull request by jeremycrisp**
Add support for pre and post conditions

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]
Mostly fine but I would like to know why some of the Attribute unsupported features in StratoX have vanished.

File: experiments/golden-results/StratoX-summary.txt [hunk 1]
```diff
@@ -73,17 +63,12 @@ Calling function: Do_Expression
 Error message: Index of string unsupported
 Nkind: N_Indexed_Component
 --
-Occurs: 34 times
```
Comment (martin-cs) [inline experiments/golden-results/StratoX-summary.txt:76]: Why has this unsupported feature vanished?

File: experiments/golden-results/StratoX-summary.txt [hunk 2]
```diff
@@ -93,20 +78,15 @@ Calling function: Get_Bound
...
 --
-Occurs: 15 times
+Occurs: 13 times
 Calling function: Process_Statement
 Error message: Unknown expression kind
 Nkind: N_Object_Renaming_Declaration
 --
 Occurs: 12 times
 Calling function: Do_Expression
-Error message: ATTRIBUTE_ROUNDING unsupported
```
Comment (martin-cs) [inline experiments/golden-results/StratoX-summary.txt:103]: Another one.

File: experiments/golden-results/StratoX-summary.txt [hunk 3]
```diff
@@ -173,11 +138,6 @@ Calling function: Process_Pragma_Declaration
 Error message: Unknown pragma: ghost
 Nkind: N_Pragma
 --
-Occurs: 4 times
```
Comment (martin-cs) [inline experiments/golden-results/StratoX-summary.txt:176]: Again, not clear what has happened to these.

File: experiments/golden-results/libkeccak-summary.txt [hunk 4]
```diff
@@ -38,25 +28,15 @@ Calling function: Do_Expression
...
 Occurs: 40 times
 Calling function: Process_Pragma_Declaration
 Error message: Unknown pragma: export_procedure
 Nkind: N_Pragma
 --
-Occurs: 34 times
-Calling function: Process_Pragma_Declaration
-Error message: Unknown pragma: ghost
-Nkind: N_Pragma
+Occurs: 36 times
```
Comment (martin-cs) [inline experiments/golden-results/libkeccak-summary.txt:36]: Why has this decreased?

File: experiments/golden-results/libkeccak-summary.txt [hunk 5]
```diff
@@ -163,6 +138,11 @@ Calling function: Do_Aggregate_Literal_Array
 Error message: Array aggregates of type string are unsupported
 Nkind: N_Aggregate
 --
+Occurs: 9 times
```
Comment (martin-cs) [inline experiments/golden-results/libkeccak-summary.txt:141]: Why are there now fewer of these?

File: experiments/golden-results/vct-summary.txt [hunk 6]
```diff
@@ -177,3 +172,13 @@ Occurs: 1 times
 +===========================GNAT BUG DETECTED==============================+
 | GNU Ada (ada2goto) Assert_Failure sinfo.adb:3253                         |
 Error detected at REDACTED
+--
+Occurs: 1 times
++===========================GNAT BUG DETECTED==============================+
+| GNU Ada (ada2goto) exception ureal_to_binary.adb:258                     |
```
Comment (martin-cs) [inline experiments/golden-results/vct-summary.txt:178]: Any idea why we're getting extra crashes?

**Review by martin-cs — CHANGES_REQUESTED** [review]
Mostly looks really good but I still don't understand why some of the seemingly unrelated unsupported feature reports are either vanishing or becoming less frequent.  Please can you look into this.  There may be an good / innocent / positive explanation.

File: experiments/golden-results/vct-summary.txt (hunk 6 above)
Comment (jeremycrisp) [reply experiments/golden-results/vct-summary.txt:178]: This is within system.ads and is caused by having a negative exponent value. Unsure as to why is is now being reported.

Comment (martin-cs) [reply experiments/golden-results/vct-summary.txt:178]: OK.

**Review by martin-cs — APPROVED** [review]

**Pull request merged**